### ISSUE TITLE
[Clang] Possible miscompilation of the destructor of a class with a virtual base class

### ISSUE THREAD

**Issue by frederick-vs-ja** [llvm/llvm-project#74282]
The following program attempts to destroy a **mediate base class subobject whose type has a virtual base class**.

(Skipping the side effects of a destructor is made well-defined via [CWG2523](https://cplusplus.github.io/CWG/issues/2523.html).)

```C++
#include <cstdio>

struct B {
    B() { std::printf("B::B %p\n", static_cast<const void*>(this)); }
    B(const B&) = delete;
    B &operator=(const B&) = delete;
    ~B() { std::printf("B::~B %p\n", static_cast<const void*>(this)); }
};

struct DX : virtual B {
    DX() { std::printf("DX::DX %p\n", static_cast<const void*>(this)); }
    DX(const DX&) = delete;
    DX &operator=(const DX&) = delete;
    ~DX() { std::printf("DX::~DX %p\n", static_cast<const void*>(this)); }
};

struct DY : virtual B {
    DY() { std::printf("DY::DY %p\n", static_cast<const void*>(this)); }
    DY(const DY&) = delete;
    DY &operator=(const DY&) = delete;
    ~DY() { std::printf("DY::~DY %p\n", static_cast<const void*>(this)); }
};

struct DZ : DX, DY {
    DZ() { std::printf("DZ::DZ %p\n", static_cast<const void*>(this)); }
    DZ(const DZ&) = delete;
    DZ &operator=(const DZ&) = delete;
    ~DZ() { std::printf("DZ::~DZ %p\n", static_cast<const void*>(this)); }
};

template<class T>
union NoDestroy {
    T val;
    NoDestroy() : val() {}
    NoDestroy(const NoDestroy&) = delete;
    NoDestroy &operator=(const NoDestroy&) = delete;
    ~NoDestroy() {}
};

int main()
{
    NoDestroy<DZ> d{};
    static_cast<DY&>(d.val).~DY();
}
```

When using Clang (same for GCC), the addresses printed after `B::B` and `B::~B` are always different ([Godbolt link](https://godbolt.org/z/q4sYoWhTx)). Clang (and GCC) seems to emit miscalculation of the offset of the `B` base class subobject in `DY::~DY` (possibly due to early setting vptr), which is likely wrong when the `DY` object is not a most derived object.

MSVC seems to calculate the offset correctly.

There doesn't seem anything in the C++ standard indicating that "destructing a mediate base class subobject whose type has a virtual base class" is undefined behavior. However, [[class.dtor]/13](https://eel.is/c++draft/class.dtor#13) says
> [...] and, if `X` is the most derived class ([class.base.init]), its destructor calls the destructors for `X`'s virtual base classes. [...]

which looks defective as it possibly requires additional metadata to tell whether a class object whose type has a virtual base class is a most derived object.

**Comment by changkhothuychung** [comment]
If this is deemed an issue, I would like to take this issue. Since I am still new to LLVM, if someone can provide the files I should look into, and some steps to begin, it would be great, thanks.

**Comment by changkhothuychung** [comment]
> Godbolt link



> The following program attempts to destroy a **mediate base class subobject whose type has a virtual base class**.
> 
> (Skipping the side effects of a destructor is made well-defined via [CWG2523](https://cplusplus.github.io/CWG/issues/2523.html).)
> 
> ```c++
> #include <cstdio>
> 
> struct B {
>     B() { std::printf("B::B %p\n", static_cast<const void*>(this)); }
>     B(const B&) = delete;
>     B &operator=(const B&) = delete;
>     ~B() { std::printf("B::~B %p\n", static_cast<const void*>(this)); }
> };
> 
> struct DX : virtual B {
>     DX() { std::printf("DX::DX %p\n", static_cast<const void*>(this)); }
>     DX(const DX&) = delete;
>     DX &operator=(const DX&) = delete;
>     ~DX() { std::printf("DX::~DX %p\n", static_cast<const void*>(this)); }
> };
> 
> struct DY : virtual B {
>     DY() { std::printf("DY::DY %p\n", static_cast<const void*>(this)); }
>     DY(const DY&) = delete;
>     DY &operator=(const DY&) = delete;
>     ~DY() { std::printf("DY::~DY %p\n", static_cast<const void*>(this)); }
> };
> 
> struct DZ : DX, DY {
>     DZ() { std::printf("DZ::DZ %p\n", static_cast<const void*>(this)); }
>     DZ(const DZ&) = delete;
>     DZ &operator=(const DZ&) = delete;
>     ~DZ() { std::printf("DZ::~DZ %p\n", static_cast<const void*>(this)); }
> };
> 
> template<class T>
> union NoDestroy {
>     T val;
>     NoDestroy() : val() {}
>     NoDestroy(const NoDestroy&) = delete;
>     NoDestroy &operator=(const NoDestroy&) = delete;
>     ~NoDestroy() {}
> };
> 
> int main()
> {
>     NoDestroy<DZ> d{};
>     static_cast<DY&>(d.val).~DY();
> }
> ```
> 
> When using Clang (same for GCC), the addresses printed after `B::B` and `B::~B` are always different ([Godbolt link](https://godbolt.org/z/q4sYoWhTx)). Clang (and GCC) seems to emit miscalculation of the offset of the `B` base class subobject in `DY::~DY` (possibly due to early setting vptr), which is likely wrong when the `DY` object is not a most derived object.
> 
> MSVC seems to calculate the offset correctly.
> 
> There doesn't seem anything in the C++ standard indicating that "destructing a mediate base class subobject whose type has a virtual base class" is undefined behavior. However, [[class.dtor]/13](https://eel.is/c++draft/class.dtor#13) says
> 
> > [...] and, if `X` is the most derived class ([class.base.init]), its destructor calls the destructors for `X`'s virtual base classes. [...]
> 
> which looks defective as it possibly requires additional metadata to tell whether a class object whose type has a virtual base class is a most derived object.

Hi, if you have sometimes, can you help me explain the following concepts? 

1) `the addresses printed after B::B and B::~B are always different ` - Do you mean the addresses of B() and ~B() are supposed to be always the same? 

2) `Clang (and GCC) seems to emit miscalculation of the offset of the B base class subobject in DY::~DY` - Can you elaborate on a little more on this how it is supposed to work? If I understand correctly, since B is a virtual base class of DY, then DY will have a vtable which has ptr to the virtual functions of B, is this correct? What miscalculation are you seeing? 

I am new to this topic so I am just trying to learn, any input is greatly appreciated!

**Comment by frederick-vs-ja** [comment]
> 1\. `the addresses printed after B::B and B::~B are always different ` - Do you mean the addresses of B() and ~B() are supposed to be always the same?

No... I meant that `B()` and `~B()` are supposed to construct and destroy the same subobject. The prited addresses are the addresses of the object argument for which ctor and dtor are called. I think I've expressed this clearly in the code. Did I miss anything?

> 2\. `Clang (and GCC) seems to emit miscalculation of the offset of the B base class subobject in DY::~DY` - Can you elaborate on a little more on this how it is supposed to work? If I understand correctly, since B is a virtual base class of DY, then DY will have a vtable which has ptr to the virtual functions of B, is this correct? What miscalculation are you seeing?

For example, in an execution the output can be:
```
B::B 0x7ffc8c676b20
DX::DX 0x7ffc8c676b20
DY::DY 0x7ffc8c676b28
DZ::DZ 0x7ffc8c676b20
DY::~DY 0x7ffc8c676b28
B::~B 0x7ffc8c676b28
```

The last line should be `B::~B 0x7ffc8c676b20` (i.e. the printed address after `B::~B` is same as that after `B::B`) in an expected result.
